pkg/healthcheck/handler.go

```diff
@@ -15,83 +15,130 @@
 package healthcheck
 
 import (
+	"context"
 	"net"
 	"net/http"
 	"strconv"
+	"sync/atomic"
 
 	"go.uber.org/zap"
 
 	"github.com/jaegertracing/jaeger/pkg/version"
 )
 
-// State represents the current health check state
-type State struct {
-	state   int
+// Status represents the state of the service.
+type Status int
+
+const (
+	// Unavailable indicates the service is not able to handle requests
+	Unavailable Status = iota
+	// Ready indicates the service is ready to handle requests
+	Ready
+	// Broken indicates that the healthcheck itself is broken, not serving HTTP
+	Broken
+)
+
+func (s Status) String() string {
+	switch s {
+	case Unavailable:
+		return "unavailable"
+	case Ready:
+		return "ready"
+	case Broken:
+		return "broken"
+	default:
+		return "unknown"
+	}
+}
+
+// HealthCheck provides an HTTP endpoint thta returns the health status of the service
+type HealthCheck struct {
+	state   int32 // atomic, keep at the top to be word-aligned
 	logger  *zap.Logger
-	onClose chan struct{}
+	mapping map[Status]int
+	server  *http.Server
 }
 
-// Serve requests on the specified port. The initial state is what's specified with the state parameter
-func Serve(state int, port int, logger *zap.Logger) (*State, error) {
-	hs, err := NewState(state, logger)
-	handler, err := NewHandler(hs)
+// Option is a functional option for passing paameters to New()
+type Option func(*HealthCheck)
+
+// Logger creates an option to set the logger. If not specified, Nop logger is used.
+func Logger(logger *zap.Logger) Option {
+	return func(hc *HealthCheck) {
+		hc.logger = logger
+	}
+}
+
+// New creates a HealthCheck with the specified initial state.
+func New(state Status, options ...Option) *HealthCheck {
+	hc := &HealthCheck{
+		state: int32(state),
+		mapping: map[Status]int{
+			Unavailable: http.StatusServiceUnavailable,
+			Ready:       http.StatusNoContent,
+		},
+	}
+	for _, option := range options {
+		option(hc)
+	}
+	if hc.logger == nil {
+		hc.logger = zap.NewNop()
+	}
+	return hc
+}
 
-	s := &http.Server{Handler: handler}
+// Serve starts HTTP server on the specified port.
+func (hc *HealthCheck) Serve(port int) (*HealthCheck, error) {
 	portStr := ":" + strconv.Itoa(port)
 	l, err := net.Listen("tcp", portStr)
 	if err != nil {
-		logger.Error("failed to listen", zap.Error(err))
+		hc.logger.Error("failed to listen", zap.Error(err))
```

Comment:
Could you add more context similar to line 99, sth like Health Check server failed to listen, or so?